[Lim Wei Liang] iP

limweiliang opened the pull request:
# Athena
> "Your mind is for having ideas, not holding them" - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Athena is your personal assistant for remembering all the things you need to do! 
It's
* text-based
* easy to learn
* ~~FAST~~*SUPER* FAST to use

To get things up and running, simply:
1. download it from releases on the right
2. double-click on it
3. add your tasks/events/deadlines
4. let it manage your tasks for you 😄 

and, it is **FREE** and **OPEN-SOURCE**!!

Features:
- [X] Task Management
- [ ] Track Upcoming Events
- [ ] Track Upcoming Deadlines

---

If you're new to Java, you can also take a look to see how basic Java features are used! For example, this is the `main` method.
```java
public static void main(String[] args) {
    new Main().run();
}
```

Commit:
Add Gradle support

limweiliang committed:
added Athena.java

limweiliang committed:
Added TrackingList.java, updated Athena to store and return input

limweiliang committed:
Added Task.java, changed TrackingList.java to store Tasks

limweiliang committed:
Updated Athena.java to support marking and unmarking of tasks

limweiliang committed:
Added 3 variants of tasks: deadlines, events and todos

limweiliang committed:
Updated Athena.java to support todos, events and deadlines

limweiliang committed:
Added automated test for commands

limweiliang committed:
Added error checking for mark and unmark commands

limweiliang committed:
added error detection for todo, deadline and event

limweiliang committed:
Added support for deleting tasks

limweiliang committed:
Add exception handling for incorrect inputs

limweiliang committed:
Add save to disk functionality for Athena

limweiliang committed:
Add load from disk functionality

limweiliang committed:
Store datetimes for event/deadline as LocalDateTime instead of String

limweiliang committed:
Merge branch 'branch-Level-7'

limweiliang committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Athena.java

limweiliang committed:
Refactor code to create separate Storage class

limweiliang committed:
Refactor code into more classes

limweiliang committed:
refactor code by dividing classes into packages

limweiliang committed:
reorganize file structure in project

limweiliang committed:
Add JUnit tests for Parser.java and Todo.java

limweiliang committed:
Add Javadoc comments for commands, exceptions and Parser.java

limweiliang committed:
Add JavaDoc comments for Main.java and Storage.java

limweiliang committed:
Refactor code to adhere to coding standards

limweiliang committed:
Add a 'find' command to Athena to search by a given phrase

limweiliang committed:
Merge branch 'branch-A-JavaDoc'

limweiliang committed:
Merge branch 'branch-A-CodingStandard'

limweiliang committed:
Merge branch 'branch-Level-9'

limweiliang committed:
Fix bug arising from merging 'branch-Level-9'

flairekq commented on src/main/java/athena/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+ * Represents a single command provided to Athena by the user. This is
+ * an umbrella class for all types of commands.
+ */
+public abstract class Command {
```
I like how you further extract the codes for the Command related classes, making the code looks clean!

flairekq commented on src/main/java/athena/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+ * Represents an unmark command given to Athena by the user. When executed, marks
+ * the task corresponding to the given task number as not done.
+ */
+public class UnmarkCommand extends Command {
```
Perhaps you can try combining Mark and Unmark command into a single command since both are quite similar to reduce duplicated codes

flairekq commented on src/main/java/athena/tasks/Event.java:
```diff
@@ -0,0 +1,36 @@
+
+// STRETCH GOAL: Add a duration for event?
+public class Event extends Task {
+    private static final DateTimeFormatter outputFormatter =
```
perhaps using uppercase and underscore to separate the words for outputFormatter e.g. OUTPUT_FORMATTER helps in conforming to the coding standard

flairekq commented on src/main/java/athena/tasks/Event.java:
```diff
@@ -0,0 +1,36 @@
+public class Event extends Task {
+    private static final DateTimeFormatter outputFormatter =
+            DateTimeFormatter.ofPattern("MMM d yyyy hh:mma");
+    private final LocalDateTime eventDate;
```
similarly for eventDate can use uppercase and underscore

flairekq reviewed (commented):
Overall I like how descriptive the header comments are for those that are written. LGTM, almost. Just a few nits to fix.

NoraYUuu commented on src/main/java/athena/Main.java:
```diff
@@ -0,0 +1,86 @@
+     * requisite objects and data structures.
+     */
+    public Main() {
+        storage = new Storage(SAVE_DIRECTORY, SAVE_FILENAME);
```
I like how you declared the file path for memory as static constants, this makes it more convenient if there is a need to change!

FYimu replied:
I like how you declared the file path for memory as static constants, this makes it more convenient if there is a need to change!

FYimu commented on src/main/java/athena/Main.java:
```diff
@@ -0,0 +1,86 @@
+            } catch (InputException e) {
+                ui.sayText(e.getMessage());
+            }
+            SaveIfTaskListModified();
```
I like how you check if the storage needs to be updated after the execution of each command, this saves the trouble of passing storage into each execute method and overriding it in all command classes!

FYimu commented on src/main/java/athena/Main.java:
```diff
@@ -0,0 +1,86 @@
+import athena.tasks.TaskList;
+import athena.ui.Ui;
+
+import java.io.IOException;
```
I like how you organized all your imports in alphabetic order. It looks clean and it's convenient to find a particular import!

FYimu commented on src/main/java/athena/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Creates a new deadline task in the given TaskList and outputs the results to the
+     * given Ui.
```
It might align better with coding standard if you write one full sentence on one line for JavaDoc 😄 (if I remember correctly from CheckStyle)

limweiliang replied:
Yes, good catch, will try to keep my summary sentences succinct!

FYimu commented on src/main/java/athena/commands/FindCommand.java:
```diff
@@ -0,0 +1,42 @@
+    public void execute(Ui ui, TaskList taskList) {
+        List<Integer> taskNumbers = taskList.getTaskNumbersContainingPhrase(searchPhrase);
+        if (taskNumbers.size() > 0) {
+            ui.sayText(String.format("Here are the tasks containing the phrase '%s'", searchPhrase));
```
You might consider use `ui.sayText(searchPhrase)` and move all String to ui class, so that when you want to change the message wording, you can find everything inside the ui class which specializes in dealing with user interface 😸

FYimu commented on src/main/java/athena/commands/TodoCommand.java:
```diff
@@ -0,0 +1,52 @@
+        if (other instanceof TodoCommand) {
+            TodoCommand otherCommand = (TodoCommand) other;
+            return this.taskName.equals(otherCommand.taskName);
+        } else {
```
A very very very very small suggestion here, this can be modified as 
```
    if (other instanceof TodoCommand) {
        TodoCommand otherCommand = (TodoCommand) other;
        return this.taskName.equals(otherCommand.taskName);
    }
    return false;
```
Looks a little bit more concise ~

FYimu commented on src/main/java/athena/exceptions/InputErrorCode.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Represents a possible type of user input error when giving commands to Athena.
+ */
+public enum InputErrorCode {
```
I like how to create an Enum class for errors! Maybe I will try in my code next time 🤔

FYimu commented on src/main/java/athena/parser/Parser.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    private static String getTaskNameFromInput(String input) throws InputException {
+        // Remove excess input arising from extra fields like /by and /at
```
I like how you commented before some code blocks to enhance readability of your code XD

FYimu commented on src/main/java/athena/tasks/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Deadline extends Task {
+    private static final DateTimeFormatter outputFormatter =
+            DateTimeFormatter.ofPattern("MMM d yyyy hh:mma");
```
ALL_CAPITAL_FORMAT for static constants like @flairekq mentioned below~

FYimu commented on src/main/java/athena/ui/Ui.java:
```diff
@@ -0,0 +1,55 @@
+        for (int taskNumber : taskNumbers) {
+            System.out.printf("%d. %s", taskNumber, taskList.getTaskString(taskNumber));
+            System.out.println();
+        }
```
What I did for this is to edit the toString() method of TaskList to print all the tasks, because I personally feel like Ui is a user interface that only handles String and the display of String, so I tried to avoid asking Ui to iterate through TaskList ;P
Mind telling me your rationale too?

limweiliang replied:
hmm, because this one was for the `find` command where I only wanted to print out specific tasks, so I thought it made more sense to keep it in the `Ui` class.

FYimu replied:
True true. I solved this problem by creating a new `TaskList` instance containing only filtered tasks and call `toString()` method on it haha

FYimu commented on src/test/java/athena/tasks/TodoTest.java:
```diff
@@ -0,0 +1,58 @@
+    @Test
+    void getIcon() {
+        assertEquals("T", markedTodo.getIcon());
+    }
```
I like how you automated your test procedures!

limweiliang replied:
by the way, not related to your comment but I think we were supposed to leave our comments as a single code review rather than individual comments, otherwise the automated system may not pick this up.

But thank you for the detailed comments! 😄

FYimu replied:
Sorries I didn't notice this until I did Week 4 quiz just now 😭 Hope you were not disturbed a lot when I made so many comments 😭 😭 😭 
Thanks for your reminder! It was an enjoyable experience reading your code!

limweiliang committed:
Change variable names to fit coding standards

limweiliang committed:
Merge branch 'add-gradle-support' of https://github.com/limweiliang/ip

limweiliang committed:
Update Gradle file to fit project

limweiliang committed:
Add CheckStyle config files and tweak code to fix CheckStyle errors

FYimu commented on src/main/java/athena/Main.java:
```diff
@@ -0,0 +1,86 @@
+import athena.parser.Parser;
+import athena.storage.Storage;
+import athena.tasks.TaskList;
+import athena.ui.Ui;
```
Good job on organizing your imports in lexicographical order!

FYimu commented on src/main/java/athena/Main.java:
```diff
@@ -0,0 +1,86 @@
+                ui.sayText(e.getMessage());
+            }
+            saveIfTaskListModified();
+        }
```
I like how you checked if it is necessary to update storage after each command, making sure that all modifications are stored even if there is a bug in the future!

FYimu reviewed (commented):
Hii me again, leaving a review for grading purpose..

Sorries for troubling you with so many comment notifications 😭

limweiliang committed:
Add basic GUI functionality to display commands and responses

limweiliang committed:
Update GUI code to use FXML

limweiliang committed:
Add styling to GUI using css

limweiliang committed:
Refactor code to output to GUI

limweiliang committed:
Merge branch 'branch-Level-10'

limweiliang committed:
Remove redundant imports and fix build.gradle

limweiliang committed:
Add assertions to ensure code integrity

Currently, the code does not make use of assertions to ensure
that assumptions hold in important methods. This makes the
program more prone to bugs.

Let's add assertions to methods that rely on assumptions
about arguments and about the contents of certain variables.

Assertions are a good choice for checking programmer assumptions.

limweiliang committed:
Refactor command classes to improve code quality

Currently, execute method in many command classes does not adhere
to SLAP, and does not make use of guard clauses.

This decreases code readability and maintainability.

Let's refactor the code so that all dialog-producing code
is handled in the Messages class instead, and guard clauses
are used when possible.

This allows for a clear separation of responsibilities amongst
classes and better adherence to SLAP.

limweiliang committed:
Refactor code to improve code quality

limweiliang committed:
Merge pull request #1 from limweiliang/branch-A-Assertions

Add assertions to ensure code integrity

limweiliang committed:
Merge branch 'master' of https://github.com/limweiliang/ip

* 'master' of https://github.com/limweiliang/ip:
  Add assertions to ensure code integrity

limweiliang committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/athena/Athena.java
#	src/main/java/athena/tasks/TaskList.java
#	src/main/java/athena/ui/Messages.java

limweiliang committed:
Merge pull request #2 from limweiliang/branch-A-CodeQuality

Improve code quality through refactoring

limweiliang committed:
Merge branch 'master' of https://github.com/limweiliang/ip

* 'master' of https://github.com/limweiliang/ip:
  Refactor code to improve code quality
  Refactor command classes to improve code quality

limweiliang committed:
Add support for getting reminders about upcoming events/deadlines

limweiliang committed:
Merge branch 'branch-B-Reminders'

limweiliang committed:
Modify GUI to show user input errors in a different colour

limweiliang committed:
Add Javadocs for Athena.java and Tasklist.java

limweiliang committed:
Add product screenshot

limweiliang committed:
Add user guide

limweiliang committed:
Update product documentation

limweiliang committed:
Set theme jekyll-theme-merlot

limweiliang committed:
Set theme jekyll-theme-slate

limweiliang committed:
Fix typo in user guide

limweiliang committed:
Merge branch 'master' of https://github.com/limweiliang/ip
